src/cargo/core/compiler/unit.rs

```diff
@@ -52,6 +52,8 @@ pub struct UnitInner {
     pub kind: CompileKind,
     /// The "mode" this unit is being compiled for. See [`CompileMode`] for more details.
     pub mode: CompileMode,
+    /// The "mode" of the root unit. Required when unit's mode is `CompileMode::RunCustomBuild`
+    pub root_mode: CompileMode,
```

Comment:
Rather than adding an additional field you don't always need, I think it might make sense to add a field to the RunCustomBuild unit:
```rust
enum CompileMode {
    RunCustomBuild { root_mode: Box<CompileMode> },
}
```

Reply:
@jyn514 that's an awesome idea, thanks!  One issue -- the enum implements `Copy` trait which cannot be auto-implemented. I wonder if i can get away with the `Arc` instead?